File: skills/sync-utils/scripts/sync_status.py
#!/usr/bin/env python3
"""
Sync Status - Zeigt Status aller Skills
"""

import os
import sys
import json
from pathlib import Path
from datetime import datetime

# Import sync functions
sys.path.append('/home/openclaw/.openclaw/workspace/scripts')
from sync_clawhub_git import get_file_hash

CLAWHUB_DIR = Path("/home/openclaw/.openclaw/workspace/skills")
GIT_DIR = Path("/home/openclaw/.openclaw/workspace/git/skills")
STATE_FILE = Path("/home/openclaw/.openclaw/workspace/db/sync_state.json")

def check_skill_status(skill_name: str) -> dict:
    """Prüft Status eines Skills"""
    clawhub_path = CLAWHUB_DIR / skill_name
    git_path = GIT_DIR / skill_name
    
    status = {
        "name": skill_name,
        "in_clawhub": clawhub_path.exists(),
        "in_git": git_path.exists(),
        "has_git_repo": (git_path / ".git").exists() if git_path.exists() else False,
        "status": "unknown",
        "last_modified": {}
    }
    
    # Status bestimmen
    if status["in_clawhub"] and not status["in_git"]:
        status["status"] = "only_clawhub"
    elif status["in_git"] and not status["in_clawhub"]:
        status["status"] = "only_git"
    elif status["in_clawhub"] and status["in_git"]:
        # Timestamps vergleichen
        try:
            clawhub_mtime = max(p.stat().st_mtime for p in clawhub_path.rglob('*') if p.is_file())
            git_mtime = max(p.stat().st_mtime for p in git_path.rglob('*') if p.is_file() and '.git' not in str(p))
            
            status["last_modified"]["clawhub"] = datetime.fromtimestamp(clawhub_mtime).strftime('%Y-%m-%d %H:%M:%S')
            status["last_modified"]["git"] = datetime.fromtimestamp(git_mtime).strftime('%Y-%m-%d %H:%M:%S')
            
            if abs(clawhub_mtime - git_mtime) < 60:
                status["status"] = "synced"
            elif clawhub_mtime > git_mtime:
                status["status"] = "clawhub_newer"
            else:
                status["status"] = "git_newer"
        except:
            status["status"] = "error"
    
    return status

def main():
    """Hauptfunktion"""
    print("=" * 80)
    print("ClawHub ↔ Git Sync Status")
    print("=" * 80)
    print(f"Zeitpunkt: {datetime.now().strftime('%Y-%m-%d %H:%M:%S')}")
    print()
    
    # Alle Skills finden
    all_skills = set()
    if CLAWHUB_DIR.exists():
        all_skills.update(d.name for d in CLAWHUB_DIR.iterdir() if d.is_dir() and not d.name.startswith('.'))
    if GIT_DIR.exists():
        all_skills.update(d.name for d in GIT_DIR.iterdir() if d.is_dir() and not d.name.startswith('.'))
    
    # Status-Kategorien
    categories = {
        "synced": [],
        "clawhub_newer": [],
        "git_newer": [],
        "only_clawhub": [],
        "only_git": [],
        "error": []
    }
    
    # Status für jeden Skill prüfen
    for skill in sorted(all_skills):
        status = check_skill_status(skill)
        categories[status["status"]].append(status)
    
    # Ausgabe
    print(f"📊 Gesamt: {len(all_skills)} Skills\n")
    
    # Synchronisiert
    if categories["synced"]:
        print(f"✅ Synchronisiert ({len(categories['synced'])})")
        for s in categories["synced"]:
            print(f"   - {s['name']}")
        print()
    
    # ClawHub neuer
    if categories["clawhub_newer"]:
        print(f"🔄 ClawHub neuer ({len(categories['clawhub_newer'])})")
        for s in categories["clawhub_newer"]:
            print(f"   - {s['name']} (ClawHub: {s['last_modified']['clawhub']})")
        print()
    
    # Git neuer
    if categories["git_newer"]:
        print(f"🔄 Git neuer ({len(categories['git_newer'])})")
        for s in categories["git_newer"]:
            print(f"   - {s['name']} (Git: {s['last_modified']['git']})")
        print()
    
    # Nur in ClawHub
    if categories["only_clawhub"]:
        print(f"📦 Nur in ClawHub ({len(categories['only_clawhub'])})")
        for s in categories["only_clawhub"]:
            print(f"   - {s['name']}")
        print()
    
    # Nur in Git
    if categories["only_git"]:
        print(f"📁 Nur in Git ({len(categories['only_git'])})")
        for s in categories["only_git"]:
            print(f"   - {s['name']}")
        print()
    
    # Fehler
    if categories["error"]:
        print(f"❌ Fehler ({len(categories['error'])})")
        for s in categories["error"]:
            print(f"   - {s['name']}")
        print()
    
    # State-File Info
    if STATE_FILE.exists():
        with open(STATE_FILE, 'r') as f:
            state = json.load(f)
        last_runs = list(state.get("last_sync", {}).keys())
        if last_runs:
            print(f"📅 Letzter automatischer Sync: {last_runs[-1]}")
    
    print("=" * 80)

if __name__ == "__main__":
    main()

File: javascript/sync_status.js
#!/usr/bin/env node
// sync_status.py — portiert nach javascript
// Quelle: python, OpenClaw@gateway1:skills/sync-utils/scripts/sync_status.py
// auch in: OpenClaw@gateway2:skills/sync-utils/scripts/sync_status.py
// Erzeugt: 2026-08-23 durch ABSTRACTIONS_MANAGER.py

/**
 * Sync Status - Zeigt Status aller Skills
 */

const fs = require('fs');
const path = require('path');
const crypto = require('crypto');

// Konstanten
const CLAWHUB_DIR = "/home/openclaw/.openclaw/workspace/skills";
const GIT_DIR = "/home/openclaw/.openclaw/workspace/git/skills";
const STATE_FILE = "/home/openclaw/.openclaw/workspace/db/sync_state.json";

/**
 * Berechnet Hash einer Datei
 */
function getFileHash(filePath) {
    const hash = crypto.createHash('sha256');
    const fileBuffer = fs.readFileSync(filePath);
    hash.update(fileBuffer);
    return hash.digest('hex');
}

/**
 * Prüft Status eines Skills
 */
function checkSkillStatus(skillName) {
    const clawhubPath = path.join(CLAWHUB_DIR, skillName);
    const gitPath = path.join(GIT_DIR, skillName);
    
    const status = {
        name: skillName,
        in_clawhub: fs.existsSync(clawhubPath),
        in_git: fs.existsSync(gitPath),
        has_git_repo: fs.existsSync(path.join(gitPath, ".git")),
        status: "unknown",
        last_modified: {}
    };
    
    // Wenn nicht in Git, dann has_git_repo auf false setzen
    if (!status.in_git) {
        status.has_git_repo = false;
    }
    
    // Status bestimmen
    if (status.in_clawhub && !status.in_git) {
        status.status = "only_clawhub";
    } else if (status.in_git && !status.in_clawhub) {
        status.status = "only_git";
    } else if (status.in_clawhub && status.in_git) {
        // Timestamps vergleichen
        try {
            // Finde alle Dateien und deren mtime in clawhub
            let clawhubFiles = [];
            function walkClawhub(dir) {
                const files = fs.readdirSync(dir);
                for (const file of files) {
                    const filePath = path.join(dir, file);
                    const stat = fs.statSync(filePath);
                    if (stat.isDirectory()) {
                        walkClawhub(filePath);
                    } else {
                        clawhubFiles.push({path: filePath, mtime: stat.mtimeMs});
                    }
                }
            }
            
            walkClawhub(clawhubPath);
            const clawhubMtimes = clawhubFiles.map(f => f.mtime);
            const clawhubMtime = Math.max(...clawhubMtimes);
            
            // Finde alle Dateien und deren mtime in git (ohne .git)
            let gitFiles = [];
            function walkGit(dir) {
                const files = fs.readdirSync(dir);
                for (const file of files) {
                    const filePath = path.join(dir, file);
                    if (filePath.includes('.git')) continue;
                    const stat = fs.statSync(filePath);
                    if (stat.isDirectory()) {
                        walkGit(filePath);
                    } else {
                        gitFiles.push({path: filePath, mtime: stat.mtimeMs});
                    }
                }
            }
            
            walkGit(gitPath);
            const gitMtimes = gitFiles.map(f => f.mtime);
            const gitMtime = Math.max(...gitMtimes);
            
            status.last_modified.clawhub = new Date(clawhubMtime).toISOString().slice(0, 19).replace('T', ' ');
            status.last_modified.git = new Date(gitMtime).toISOString().slice(0, 19).replace('T', ' ');
            
            if (Math.abs(clawhubMtime - gitMtime) < 60000) { // 60 Sekunden in ms
                status.status = "synced";
            } else if (clawhubMtime > gitMtime) {
                status.status = "clawhub_newer";
            } else {
                status.status = "git_newer";
            }
        } catch (error) {
            status.status = "error";
        }
    }
    
    return status;
}

/**
 * Hauptfunktion
 */
function main() {
    console.log("=" .repeat(80));
    console.log("ClawHub ↔ Git Sync Status");
    console.log("=" .repeat(80));
    console.log(`Zeitpunkt: ${new Date().toISOString().slice(0, 19).replace('T', ' ')}`);
    console.log();
    
    // Alle Skills finden
    let allSkills = new Set();
    
    if (fs.existsSync(CLAWHUB_DIR)) {
        const clawhubDirs = fs.readdirSync(CLAWHUB_DIR)
            .filter(d => fs.statSync(path.join(CLAWHUB_DIR, d)).isDirectory() && !d.startsWith('.'));
        clawhubDirs.forEach(d => allSkills.add(d));
    }
    
    if (fs.existsSync(GIT_DIR)) {
        const gitDirs = fs.readdirSync(GIT_DIR)
            .filter(d => fs.statSync(path.join(GIT_DIR, d)).isDirectory() && !d.startsWith('.'));
        gitDirs.forEach(d => allSkills.add(d));
    }
    
    allSkills = Array.from(allSkills);
    
    // Status-Kategorien
    const categories = {
        "synced": [],
        "clawhub_newer": [],
        "git_newer": [],
        "only_clawhub": [],
        "only_git": [],
        "error": []
    };
    
    // Status für jeden Skill prüfen
    for (const skill of allSkills.sort()) {
        const status = checkSkillStatus(skill);
        categories[status.status].push(status);
    }
    
    // Ausgabe
    console.log(`📊 Gesamt: ${allSkills.length} Skills\n`);
    
    // Synchronisiert
    if (categories.synced.length > 0) {
        console.log(`✅ Synchronisiert (${categories.synced.length})`);
        for (const s of categories.synced) {
            console.log(`   - ${s.name}`);
        }
        console.log();
    }
    
    // ClawHub neuer
    if (categories.clawhub_newer.length > 0) {
        console.log(`🔄 ClawHub neuer (${categories.clawhub_newer.length})`);
        for (const s of categories.clawhub_newer) {
            console.log(`   - ${s.name} (ClawHub: ${s.last_modified.clawhub})`);
        }
        console.log();
    }
    
    // Git neuer
    if (categories.git_newer.length > 0) {
        console.log(`🔄 Git neuer (${categories.git_newer.length})`);
        for (const s of categories.git_newer) {
            console.log(`   - ${s.name} (Git: ${s.last_modified.git})`);
        }
        console.log();
    }
    
    // Nur in ClawHub
    if (categories.only_clawhub.length > 0) {
        console.log(`📦 Nur in ClawHub (${categories.only_clawhub.length})`);
        for (const s of categories.only_clawhub) {
            console.log(`   - ${s.name}`);
        }
        console.log();
    }
    
    // Nur in Git
    if (categories.only_git.length > 0) {
        console.log(`📁 Nur in Git (${categories.only_git.length})`);
        for (const s of categories.only_git) {
            console.log(`   - ${s.name}`);
        }
        console.log();
    }
    
    // Fehler
    if (categories.error.length > 0) {
        console.log(`❌ Fehler (${categories.error.length})`);
        for (const s of categories.error) {
            console.log(`   - ${s.name}`);
        }
        console.log();
    }
    
    // State-File Info
    if (fs.existsSync(STATE_FILE)) {
        const stateContent = fs.readFileSync(STATE_FILE, 'utf8');
        const state = JSON.parse(stateContent);
        const lastRuns = Object.keys(state.last_sync || {});
        if (lastRuns.length > 0) {
            console.log(`📅 Letzter automatischer Sync: ${lastRuns[lastRuns.length - 1]}`);
        }
    }
    
    console.log("=" .repeat(80));
}

main();
